cardano-integration/src/Test/Integration/Cardano/Local.hs

```diff
@@ -0,0 +1,444 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE StrictData #-}
+
+module Test.Integration.Cardano.Local
+  ( LocalTestnet(..)
+  , LocalTestnetOptions(..)
+  , Network(..)
+  , defaultOptions
+  , withLocalTestnet
+  , withLocalTestnet'
+  ) where
+
+import Control.Concurrent (threadDelay)
+import Control.Concurrent.Async (mapConcurrently_, race_)
+import Control.Exception (Exception(displayException), throw)
+import Control.Exception.Lifted (SomeException, catch)
+import Control.Monad.Base (MonadBase)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.Control (MonadBaseControl)
+import Control.Monad.Trans.Resource (MonadResource, MonadThrow(throwM), MonadUnliftIO)
+import Data.Aeson (ToJSON, object, toJSON, (.=))
+import qualified Data.Aeson.KeyMap as KM
+import Data.Functor (void, (<&>))
+import Data.List (isInfixOf)
+import Data.Maybe (fromJust)
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import Data.Time (UTCTime, addUTCTime, getCurrentTime, nominalDiffTimeToSeconds, secondsToNominalDiffTime)
+import Data.Time.Clock.POSIX (utcTimeToPOSIXSeconds)
+import qualified Data.Yaml.Aeson as YAML
+import GHC.Generics (Generic)
+import System.FilePath ((</>))
+import System.IO (Handle, IOMode(..), hClose, openFile)
+import System.Process (CreateProcess(..), ProcessHandle, StdStream(..), cleanupProcess, createProcess, proc)
+import System.Random (randomRIO)
+import Test.Integration.Cardano.Process (execCli_)
+import Test.Integration.Workspace
+  ( Workspace(..)
+  , copyToWorkspace
+  , createWorkspace
+  , createWorkspaceDir
+  , moveToWorkspace
+  , resolveWorkspacePath
+  , rewriteJSONFile
+  , rewriteYAMLFile
+  , writeWorkspaceFileJSON
+  )
+import qualified Test.Integration.Workspace as W
+import Text.Printf (printf)
+import UnliftIO.Exception (catchAny)
+import UnliftIO.Resource (allocate, runResourceT, unprotect)
+
+data LocalTestnetOptions = LocalTestnetOptions
+  { slotDuration :: Int
+  , securityParam :: Int
+  , numSpoNodes :: Int
+  , numDelegators :: Int
+  , numWallets :: Int
+  , totalBalance :: Int
+  } deriving (Show, Eq, Ord)
+
+data LocalTestnet = LocalTestnet
```

Comment:
This type is also consistent with a public testnet if we allow `delegators` and `spoNodes` to be empty. I'm not sure if we should rename this to be more general vs eventually creating a type for public testnets.

Reply:
I don't really know right now - we can always refactor when we add the public testnet support. Things will be clearer then.